ion-schema/src/model/schema.rs

```diff
@@ -0,0 +1,236 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::model::schema::schema_doc_builder_state::{BeforeHeader, HasFooter, Types};
+use crate::model::{SchemaHeader, TypeDefinition, VersionedTypeDefinition};
+use crate::{IslVersion, Versioned};
+use ion_rs::Element;
+use std::collections::HashMap;
+use std::marker::PhantomData;
+
+#[derive(Debug, Clone, PartialEq)]
+pub enum SchemaItem {
+    Header(SchemaHeader),
+    Type(String, TypeDefinition),
+    OpenContent(Element),
+    Footer(SchemaFooter),
+}
+impl From<SchemaHeader> for SchemaItem {
+    fn from(value: SchemaHeader) -> Self {
+        SchemaItem::Header(value)
+    }
+}
+impl From<SchemaFooter> for SchemaItem {
+    fn from(value: SchemaFooter) -> Self {
+        SchemaItem::Footer(value)
+    }
+}
+impl From<Element> for SchemaItem {
+    fn from(value: Element) -> Self {
+        SchemaItem::OpenContent(value)
+    }
+}
+
+/// Represents an Ion Schema document.
+///
+/// A schema is a collection of types that can be used to constrain the Ion data model.
+#[derive(Debug, Clone, PartialEq)]
+pub struct SchemaDocument {
+    /// The version in which this schema was created, but not necessarily the only version for which it is valid.
+    isl_version: (u8, u8),
+    items: Vec<SchemaItem>,
```

Comment:
Why is this a `Vec`? Wouldn't this mean we can add multiple headers and footers?

Reply:
It's a `Vec` so that we can preserve the relative order of the schema items (particularly the open content). The `SchemaDocumentBuilder` uses the type-state pattern to ensure that no more than one header and one footer are present, and that they occur in the correct order relative to the type definitions.